### PR TITLE
ARROW-12067: [Python][Doc] Document pyarrow_(un)wrap_scalar

### PR DESCRIPTION
**Pull request by pitrou**
Also other minor doc improvements, and add a test.

### CODE REVIEW

**Comment by pitrou** [comment]
@jorisvandenbossche Do you have an opinion on this?

**Comment by pitrou** [comment]
@jorisvandenbossche Ping.

File: docs/source/python/extending.rst [hunk 1]
```diff
@@ -71,157 +74,184 @@ pyarrow provides the following functions to go back and forth between
...
-occurred.  If successful, *out* is guaranteed to be non-NULL.
+Arrow C++ API pointer, and return it as a :class:`Result` object.  An error
+may be returned if the input object doesn't have the expected type.
+
+.. function:: Result<std::shared_ptr<Array>> arrow::py::unwrap_array(PyObject* obj)
+
+   Unwrap and return the Arrow C++ :class:`Array` pointer from *obj*.
 
-.. function:: Status unwrap_array(PyObject* obj, std::shared_ptr<Array>* out)
+.. function:: Result<std::shared_ptr<RecordBatch>> arrow::py::unwrap_batch(PyObject* obj)
```
Comment (jorisvandenbossche) [inline docs/source/python/extending.rst:151]: Just to understand the change: this was already changed before in C++, and the documentation was just outdated?

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

File: docs/source/python/extending.rst (hunk 1 above)
Comment (pitrou) [reply docs/source/python/extending.rst:151]: Right.

**Pull request closed without merging**